lib/listener.py

```diff
@@ -0,0 +1,153 @@
+import subprocess
+import requests
+import os
+import time
+import json
+
+from lib.debug import *
+
+from lib import settings
+import logging
+log = logging.getLogger(settings.LOGGER)
+
+
+# For use with any listener based application (Android only)
+class Listener:
+    def __init__(self, target_device, local_port, apk_path, activity, reinstall):
+
+        self.target_device = target_device
+
+        self.url = ""
+        self.install_and_launch_app(target_device, local_port, apk_path, activity, reinstall)
+
+        if self.is_emulator(target_device):
+            self.url = "http://{}:{}".format(self.get_host_ip(), local_port)
+        else:
+            self.url = "http://{}:{}".format(self.get_device_ip(target_device), 5984)
+
+        log.info("Listener running at {} ...".format(self.url))
+
+    def install_and_launch_app(self, target_device, local_port, apk_path, activity, reinstall):
+        if reinstall:
+            monkey_output = subprocess.check_output([
+                "monkeyrunner",
+                "utilities/monkeyrunner.py",
+                "--target-device={}".format(target_device),
+                "--local-port={}".format(local_port),
+                "--apk-path={}".format(apk_path),
+                "--activity={}".format(activity),
+                "--reinstall"
+            ])
+        else:
+            monkey_output = subprocess.check_output([
+                "monkeyrunner",
+                "utilities/monkeyrunner.py",
+                "--target-device={}".format(target_device),
+                "--local-port={}".format(local_port),
+                "--apk-path={}".format(apk_path),
+                "--activity={}".format(activity),
+            ])
+        log.info("OUTPUT: {}".format(monkey_output))
+
+    def is_emulator(self, target_device):
+        return target_device.startswith("emulator")
+
+    def get_host_ip(self):
+        cmd_output = subprocess.check_output("ifconfig")
+        en0_section = cmd_output.split("\n")[11]
+        full_ip = en0_section.split()[1]
+        ip = full_ip.split("/")[0]
+        return ip
+
+    def get_device_ip(self, target_device):
+        log.info("Getting Device ip ...")
+        result = subprocess.check_output(["adb", "-s", "{}".format(target_device), "shell", "netcfg"])
+        log.info("RESULT: {}".format(result))
+        ip_line = result.split('\n')[0]
+        ip = ip_line.split()[2]
+        ip = ip.split("/")[0]
+        return ip
+
+    def verify_launched(self):
+        r = requests.get(self.url)
+        log.info("GET {} ".format(r.url))
+        r.raise_for_status()
+        log.info("{}".format(r.text))
```

Comment:
This should be moved up before raise_for_status(). If there is an exception then this will never get printed.